clang/lib/CIR/CodeGen/CIRGenStmt.cpp

```diff
@@ -1157,3 +1157,45 @@ void CIRGenFunction::emitReturnOfRValue(mlir::Location loc, RValue RV,
   }
   emitBranchThroughCleanup(loc, ReturnBlock());
 }
+
+void CIRGenFunction::buildBlock(mlir::Block *bb, bool isFinished) {
+  mlir::Block *curBB = getBuilder().getInsertionBlock();
+
+  // Fall out of the current block (if necessary).
+  buildBranch(mlir::UnknownLoc::get(&CGM.getMLIRContext()), bb);
+
+  if (isFinished && bb->use_empty()) {
+    llvm_unreachable("NYI");
+  }
+
+  // MLIR blocks are always added to a Region by default, but based on the use
+  // case here it might have been the wrong region.
+  auto curFn = cast<cir::FuncOp>(CurFn);
+  curFn.getBlocks().remove(bb);
+
+  // Place the block after the current block, if possible, or else at the end of
+  // the function.
+  if (curBB && curBB->getParent())
+    curFn.getBlocks().insert(std::next(curBB->getIterator()), bb);
+  else
+    curFn.push_back(bb);
+
+  getBuilder().setInsertionPoint(bb, bb->end());
+}
+
+void CIRGenFunction::buildBranch(mlir::Location loc, mlir::Block *target) {
```

Comment:
We shouldn't be dealing with branches at this level, we usually use `cir::IfOp` and other higher level constructs. Not to mention it's way easier to reason about emitted CIR.

The only place we currently generate branches in this phase is for returns and cleanups, which will change very soon - it makes everything way more complicated. Later passes will look into `cir::IfOp` and break them down into basic blocks.

Reply:
Yup, see other comment.